junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestInstanceFactory.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2015-2018 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.api.extension;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import org.apiguardian.api.API;
+import org.junit.jupiter.api.Nested;
+
+/**
+ * {@code TestInstanceFactory} defines the API for {@link Extension
+ * Extensions} that wish to <em>create</em> test instances.
+ *
+ * <p>Common use cases include creating test instances with special construction
+ * requirements or acquiring the test from a dependency injection framework.
+ *
+ * <p>Extensions that implement {@code TestInstanceFactory} must be
+ * registered at the class level.
+ *
```

Comment:
The class-level JavaDoc for `TestInstanceFactory` should also mention that only one such extension may be registered per test class or test class hierarchy, analogous to the documentation in the User Guide.